### ISSUE TITLE
Root privilege escalation via insecure executable/config permissions

### ISSUE THREAD

**Issue by orlitzky** [NagiosEnterprises/nagioscore#424]
This is still a work-in-progress; our Nagios installation is pretty simple so I'll have to do some testing to make sure none of my proposed solutions break things.

**Problem**

When Nagios is configured, the user and group under which it will run are chosen:

```
  --with-nagios-user=<user>
                          sets user name to run nagios
  --with-nagios-group=<grp>
                          sets group name to run nagios
```

Both default to "nagios" -- the following is from `configure.ac`:

```
AC_ARG_WITH(nagios_user,AC_HELP_STRING([--with-nagios-user=<user>],[sets user name to run nagios]),nagios_user=$withval,nagios_user=nagios)
AC_ARG_WITH(nagios_group,AC_HELP_STRING([--with-nagios-group=<grp>],[sets group name to run nagios]),nagios_grp=$withval,nagios_grp=nagios)
```

Immediately after those two lines, this appears:

```
INSTALL_OPTS="-o $nagios_user -g $nagios_grp"
AC_SUBST(INSTALL_OPTS)
```

Those `INSTALL_OPTS` are then used in every invocation of the `install` command during installation.

This can lead to a root exploit in at least two ways. The fundamental problem is that the nagios daemon is intended to be launched as root, but the daemon itself and some critical configuration files are writable (in an exploitable way) by the restricted nagios user.

1. The binary for the daemon itself, i.e. `/usr/sbin/nagios` is owned by the `$nagios_user`, but will be run as root. The `$nagios_user` can overwrite that binary with whatever he wants, and root will run it.

2. The configuration file `nagios.cfg` specifies the user and group that nagios will run as, but is owned by that same `$nagios_user`. He can edit `nagios.cfg` himself, and set `nagios_user=root`, `nagios_group=root`. The command definition file `objects/commands.cfg` is also owned by `$nagios_user`, and he can place whatever commands in there he likes. Considering that the daemon will run as root the next time it is launched, this will also let the `$nagios_user` gain root eventually.

**Resolution**

At the very least, `nagios.cfg` and the executables must not be writable by anyone other than root. To ensure that the executables aren't owned by the nagios user, it suffices to set `INSTALL_OPTS=""` in `base/Makefile.in` and `cgi/Makefile.in`.

For the configuration file, I'm not so sure: is there a reason for the other configs (besides `nagios.cfg`) to be owned by the `$nagios_user` and `$nagios_grp`?

**Comment by orlitzky** [comment]
I'm going to commit a masked version to gentoo and ask for help testing. My own limited testing has shown that most things continue to work if I run the following makefile targets with empty `INSTALL_OPTS`:

* (base) install-unstripped
* install-config
* (cgi) install-unstripped
* install-html / install-classicui

For us, that's everything that we run _except_ `install-basic`. The `INSTALL_OPTS` actually are used in `install-basic`, but even there, a little over-zealously: they give ownership of the libexecdir to the nagios user.

After we run all of those install targets, there are a few things to fix up:

* The `nagios` (and `nagiostats`?) executable needs to be made mode 775 (currently: 770) to prevent an access denied error.
* The owner of the libexecdir needs to be set back to `root:root`.

Aside from that, I was able to start up the daemon and run the web interface.

Back to testing...

**Comment by hedenface** [comment]
The only reason that's standing out at the moment for the configs to be owned by the specified user and group is that a lot of systems use those particular users/groups as the users and groups who own the configs! In the case of the group - you give specific users group access and then they are your monitoring team - they do not need root in order to update the configuration. 

Also, there are a few third party utilities that rely on the configs being writable by the user they run as. It is not uncommon to add the apache user to the nagios group to allow another web UI to adjust the configuration.

Regardless - I look forward to your test results. If you'd post a link to the source here I wouldn't mind taking a look as well.

**Comment by orlitzky** [comment]
Everything works fine on our installation. I removed `INSTALL_OPTS` from every `make` invocation _except_ `make install-basic`. After `install-basic` runs, I've changed the ownership of the plugin directory back to `root:root`.

I haven't modified the source to accomplish any of that -- I'm just running `INSTALL_OPTS="" make ...` in our ebuild.

To have non-root users update the configuration is a legitimate use case, but the group that nagios runs as must be different from the group that can update the configuration. Otherwise, "nagios" can just upgrade himself to root. And I don't think there's a legitimate case to have the nagios _user_ own anything at all except the stuff he creates under e.g. `/var/lib/nagios`.

Our Gentoo bug is here:

https://bugs.gentoo.org/show_bug.cgi?id=629380

Since everyone is running nagios as `nagios:nagios` already, the least-intrusive way to fix this is probably going to involve a new group for the config files  (we don't want to change the meaning of the "nagios" user and group on people):

1. Figure out where we can use empty `INSTALL_OPTS` in the Makefile to prevent the nagios runtime user and group from owning anything that he doesn't need to write to while running.
2. Add and document a new ./configure option for`--with-config-group`.
3. Update the makefile to set the group on the configs to the value of `$with_config_group`. Then if anyone needs to be able modify the configuration, he can safely be added to this group.
4. Throw an error if anyone tries to use the same group for `--with-config-group` and `--with-nagios-group`.
5. Would be nice: throw an error if the `--with-nagios-user` is a member of the `--with-config-group`, or simply throw an error during startup if the nagios user can write to the config file. 

Basically I would try to fix the immediate problem in (1), and then work on adding back the ability to have a "configuration editor" group in (2) through (5).

**Comment by orlitzky** [comment]
> In the case of the group - you give specific users group access and then they are your monitoring team - they do not need root in order to update the configuration.

I was just thinking about this in the shower (where all good work gets done) and even in the intended use case, this allows anyone in the `nagios` group to gain root. If your web server can edit the nagios config files, he can set `nagios_user=root` and then place whatever command he wants in one of the service checks. Then any web UI exploit becomes a root exploit on the machine, and the same thing goes for any other member of the `nagios` group. If you want to let a group of non-root users edit the configs, you can't do it safely, because they can always gain root in that manner.

So, I amend my previous comment -- I don't know what to do about that. Would it be safe if `nagios.cfg` remained `root:root` but if the other configs were editable?

**Comment by orlitzky** [comment]
> Would it be safe if nagios.cfg remained root:root but if the other configs were editable?

The obvious problem with that solution doesn't seem to be a problem: if I try to set `nagios_user` or `nagios_group `in one of the files under `cfg_dir`, nagios tells me it's an invalid statement. So maybe that plan above can still work so long as we ensure that `nagios.cfg` is `root:root`.

**Comment by box293** [comment]
There are use cases like in Nagios XI where the Core Configuration Manager allows you to edit `nagios.cfg` and hence apache would need to save changes to the file.

**Comment by orlitzky** [comment]
@box293 

crap =(

I think it's pretty clear now how any non-root user who can edit `nagios.cfg` can easily gain root. I'm not familiar with Nagios XI, but someone who is will have to evaluate the impact of this on it and similar products that expect to be able to write to `nagios.cfg`.

* Can they live without that ability?
* Should we give up the charade, and ask those people to run their web servers as root if they require that ability?
* Should the `nagios_user` and `nagios_group` parameters be removed from the configuration file, and made command-line-only options? If we did the same thing with the `lock_file` parameter and everything else that gets used while the daemon is still root, then we could make it so that `nagios.cfg` is read after dropping privileges. Then it doesn't matter who can edit the file.

I know the first two are bad, and the third is a lot of work -- I'm just throwing stuff out as it comes to mind.

**Comment by hedenface** [comment]
We talked about this internally a bit today, and quite frankly I think your
third point is the one we landed on. Also, having a completely separate
group for config sounds reasonable, and it wouldn't be too terribly
difficult to port to XI once it's complete in core.

**Comment by ZZS2017** [comment]
Now. It's fixed?

**Comment by rakesh-patnaik** [comment]
I am wondering if this is a issue in my install. I have nagios3 running as nagios user. Particularly this statement "**The fundamental problem is that the nagios daemon is intended to be launched as root**" - this is not true in my install of nagios core 3.5.1 . Can someone please explain the above quoted statement?

here is a output of the nagios process running in the system:
---------------------------
# ps -ef | grep nagios3
nagios   20342     1  1 10:25 ?        00:00:00 /usr/sbin/nagios3 -d /etc/nagios3/nagios.cfg
---------------------------

**Comment by orlitzky** [comment]
@rakesh-patnaik your `ps` command shows that nagios is running as the "nagios" user, but doesn't say who the command was originally launched as. It was probably launched as root, and then later dropped privileges to the "nagios" user. How do you start the daemon?

**Comment by rakesh-patnaik** [comment]
we use the SysV init to start nagios3. 
sudo -u nagios bash -c "service nagios3 start"
we have automation tools that perform the above start procedure.

**Comment by orlitzky** [comment]
@rakesh-patnaik 

> sudo -u nagios bash -c "service nagios3 start"

Then you're safe from this, since you're using sudo to drop privileges before the daemon even starts. Just be sure you never execute `/usr/sbin/nagios3` as root!

Most people would just add the "nagios3" service to the default run level, and that's who this vulnerability affects.

**Comment by exolain** [comment]
Is there going to be a new nagios version once this issue is closed?

**Comment by hedenface** [comment]
Yes. 4.4.0 contains the fix. It was actually going to be released yesterday, but I've gotten tied up in other things. I'm anticipating releasing it Monday, as long as there are no blocking breakages discovered from now until then :)

**Comment by hedenface** [comment]
Nope. Ignore me. For some reason I thought this was a different issue. I'll take another look at this, as it's fallen off of the radar for a bit.

**Comment by hedenface** [comment]
Okay, so re-reading it, I have a bit of information regarding the issue.

We'll add a configure group (which will likely **default** to the specified nagios group [so as to avoid breaking backwards compatibility where possible]) and a configure user - but this won't be done until a major release. I'll add it to Core 5 for now, as it isn't so much work as to not make it in that release.

There have already been quite a lot of "breaking changes" released in in the last two enhancement/minor versions, I'd like that to stop where possible. By allowing configure flags to overwrite default behavior, the packagers for different distributions can use them as they see fit.

@orlitzky How does this sound as an actual achievable goal?

**Comment by orlitzky** [comment]
The timeline is fine, but the devil is in the details with regards to the implementation. Adding another user/group doesn't help if they can still edit the configuration file to say e.g. `nagios_user = root` and then place bad commands in there. The plan from https://github.com/NagiosEnterprises/nagioscore/issues/424#issuecomment-328683681 should work and I'm happy to look at the changes once you have something concrete. I think it's a big enough change to go in 5.0.0.

**Comment by OS-WS** [comment]
Hi, was this issue ever addressed? Is this CVE fixed?

Thanks in advance!
